Running API tests often logs "Error while launching com.apple.WebKit.Networking: Connection invalid"

cdumez opened the pull request:
#### 3d791e9725a598f909da5b285740d2ac468df0b9
<pre>
Running API tests often logs &quot;Error while launching com.apple.WebKit.Networking: Connection invalid&quot;
<a href="https://bugs.webkit.org/show_bug.cgi?id=256278">https://bugs.webkit.org/show_bug.cgi?id=256278</a>
rdar://108866781

Reviewed by Alex Christensen, Sihui Liu and Per Arne Vollan.

Go via the WebsiteDataStore to access the NetworkProcessProxy, like we usually
do now that the NetworkProcessProxy is owned by the WebsiteDataStore. The code
was calling `NetworkProcessProxy::ensureDefaultNetworkProcess()` directly,
which would construct a NetworkProcessProxy and destroy it right away since
it was not holding on to the RefPtr returned by ensureDefaultNetworkProcess().

In the case where we don&apos;t have a data store (prewarmed WebProcesss case), we
stash the NetworkProcess in a RefPtr data member to keep it alive until a
website data store gets assigned (at which point we transfer ownership to the
data store).

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):

Canonical link: <a href="https://commits.webkit.org/263698@main">https://commits.webkit.org/263698@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3a3e882338e13f14c26947ff4e75393433711245

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5461 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5605 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5798 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7006 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5483 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5832 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5580 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6514 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5563 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5621 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7037 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3099 "Passed tests") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/12047 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4965 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6709 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5416 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/4444 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4874 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1309 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8978 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5237 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai reviewed (approved):
r=me

pvollan reviewed (approved):
R=me.

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1198,14 +1205,15 @@ void WebProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::Connect
-                return;
-            NetworkProcessProxy::ensureDefaultNetworkProcess()->sendXPCEndpointToProcess(*weakThis);
-        });
+        // Prewarmed WebProcesses don't have a data store but still need a network process to launch properly
```
WebProcesses => web processes (to be aligned with "network process")

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -1198,14 +1205,15 @@ void WebProcessProxy::didFinishLaunching(ProcessLauncher* launcher, IPC::Connect
+        // Prewarmed WebProcesses don't have a data store but still need a network process to launch properly
+        // because the network process needs to send it the launch services database. Since the data store
+        // normally keeps the network process alive, we stash it in m_networkProcessToKeepAliveUntilDataStoreIsCreated
+        // until the prewarmed WebProcess gets assigned a data store.
```
WebProcess => web process

szewai reviewed (approved):
r=me

cdumez committed:
Running API tests often logs "Error while launching com.apple.WebKit.Networking: Connection invalid"
https://bugs.webkit.org/show_bug.cgi?id=256278
rdar://108866781

Reviewed by Alex Christensen, Sihui Liu and Per Arne Vollan.

Go via the WebsiteDataStore to access the NetworkProcessProxy, like we usually
do now that the NetworkProcessProxy is owned by the WebsiteDataStore. The code
was calling `NetworkProcessProxy::ensureDefaultNetworkProcess()` directly,
which would construct a NetworkProcessProxy and destroy it right away since
it was not holding on to the RefPtr returned by ensureDefaultNetworkProcess().

In the case where we don't have a data store (prewarmed WebProcesss case), we
stash the NetworkProcess in a RefPtr data member to keep it alive until a
website data store gets assigned (at which point we transfer ownership to the
data store).

* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::didFinishLaunching):

Canonical link: https://commits.webkit.org/263698@main
